README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
graph3d Customize 3D graphs in R
================================

Installation
------------

``` r
if(!require(devtools)) install.packages("devtools")
devtools::install_github("kassambara/graph3d")
```

Geting started
--------------

### Quick RGL scater plot

``` r
require(rgl)
#> Loading required package: rgl
data(iris)
x <- iris$Sepal.Length
y <- iris$Petal.Length
z <- iris$Sepal.Width
```

``` r
rgl_scatter(iris, groups = iris$Species, add.ellipse = TRUE,
            data.scale = TRUE)
```

![](README-unnamed-chunk-6-1.png)

Remove the bounding box:

``` r
rgl_scatter(iris, groups = iris$Species, show.bbox = FALSE)
```

![](README-unnamed-chunk-7-1.png)

Add regression planes

``` r
rgl_scatter(iris, groups = iris$Species, show.bbox = FALSE, add.reg.planes = TRUE)
```

![](README-unnamed-chunk-8-1.png)

### Build customized RGL plots

``` r
rgl_init()
rgl.spheres(x, y, z, r = 0.2, color = get_colors(iris$Species))
rgl_add_axes(x, y, z, show.bbox = TRUE)
rgl_add_ellipses(x, y, z, groups = iris$Species)
aspect3d(1,1,1)
```

![](README-unnamed-chunk-9-1.png)

### Add grids to a scatterplot3d

The function **s3d\_addgrids()** can be used as follow:

``` r
library("scatterplot3d")

data(iris)

# Basic 3d graphics
scatterplot3d(iris[,1:3])
```

![](README-add-grids-1.png)

``` r
# Remove box and add grids
scatterplot3d(iris[, 1:3], pch = 16, grid = TRUE, box = FALSE)
s3d_addgrids(iris[, 1:3], grid = c("xy", "xz", "yz"))
```

![](README-add-grids-2.png)

### 3D histogram distribution

``` r
# quakes dataset
data("quakes")
dist3d(quakes$long, quakes$lat, colvar=quakes$depth,
             breaks = 30, col.bar = "steelblue")
```

![](README-3d-histogram-distribution-1.png)

``` r
# Iris dataset
dist3d(iris$Sepal.Length, iris$Petal.Width, col.bar = "steelblue", 
             colvar=as.numeric(iris$Species))
```

![](README-3d-histogram-distribution-2.png)


NEWS.md
# graph3d 1.0.0

## Features
    
- s3d_addgrids(): Adds grids to a scatterplot3d
- dist3d(): Draws a 3D histogram distribution of two variables
- rgl_init(): Creates a new RGL device if requested or if there is no opened device
- RGL scatter plot. 
    - rgl_scatter(): quick scatter plot using rgl
    - rgl_add_axes(): Add axes to an RGL plot
    - get_colors(): Get colors for the different levels of a factor variable.
    - rgl_add_ellipses(): Add an ellipse of concentration.
    - rgl_add_linear_planes(): Add regression planes
    - rgl_group_labels(): Show the label of each group




R/dist3d.R
#' 3D histogram distribution of two variables
#' @description Draw a 3D histogram distribution of points.
#' @param x,y vectors with x and y values.
#' @param break.func a function to compute the vector of breakpoints (see ?hist).
#' @param breaks an integer specifying the number of breaks.
#' @param col.bar color for bars.
#' @param colvar the variable used for coloring points
#' @param col.point Color palette to be used for the colvar variable.
#' If col is NULL and colvar is specified, then a red-yellow-blue colorscheme (jet.col) will be used.
#' If col is NULL and colvar is not specified, then col will be grey.
#' @param clab Only if colkey = TRUE, the label to be written on top of the color key.
#' @param theta,phi The angles defining the viewing direction
#' @param ... other arguments to be passed to plot3D::hist3D()
#' @seealso \code{\link[plot3D]{hist3D}}, \code{\link[graphics]{hist}}
#' @examples
#' data(quakes)
#' dist3d(quakes$long, quakes$lat, colvar=quakes$depth,
#'    breaks =30)
#' @export
dist3d <- function(x, y, break.func = c("Sturges", "scott", "FD"), breaks = NULL,
                        col.bar = "white",
                        colvar = NULL, col.point = plot3D::gg.col(100),
                        clab=NULL, phi = 5, theta = 25, ...){

  # Compute the number of classes for a histogram
  break.func <- break.func [1]
  if(is.null(breaks)){
    x.breaks <- switch(break.func,
                       Sturges = grDevices::nclass.Sturges(x),
                       scott = grDevices::nclass.scott(x),
                       FD = grDevices::nclass.FD(x))
    y.breaks <- switch(break.func,
                       Sturges = grDevices::nclass.Sturges(y),
                       scott = grDevices::nclass.scott(y),
                       FD = grDevices::nclass.FD(y))
  } else x.breaks <- y.breaks <- breaks

  # Cut x and y variables in bins for counting
  x.bin <- seq(min(x), max(x), length.out = x.breaks)
  y.bin <- seq(min(y), max(y), length.out = y.breaks)
  xy <- table(cut(x, x.bin), cut(y, y.bin))
  z <- xy

  xmid <- 0.5*(x.bin[-1] + x.bin[-length(x.bin)])
  ymid <- 0.5*(y.bin[-1] + y.bin[-length(y.bin)])

  oldmar <- par("mar")
  par (mar = par("mar") + c(0, 0, 0, 2))
  plot3D::hist3D(x = xmid, y = ymid, z = xy, ...,
         zlim = c(-max(z)/2, max(z)), zlab = "counts", bty= "g",
         phi = phi, theta = theta,
         shade = 0.2, col = col.bar, border = "black",
         d = 1, ticktype = "detailed")

  plot3D::scatter3D(x, y,
            z = rep(-max(z)/2, length.out = length(x)),
            colvar = colvar, col = col.point,
            add = TRUE, pch = 18, clab = clab,
            colkey = list(length = 0.5, width = 0.5,
                          dist = 0.05, cex.axis = 0.8, cex.clab = 0.8)
  )
  par(mar = oldmar)

}


R/rgl_init.R
#' Initialize RGL device
#'
#' @description Creates a new RGL device if requested or if there
#'  is no opened device
#'
#' @param new.device a logical value. If TRUE, creates a new RGL device
#' @param bg the background color of the device
#' @param width the width of the device
#' @examples
#' \donttest{
#' # Load dataset
#' data(iris)
#' x <- iris$Sepal.Length
#' y <- iris$Petal.Length
#' z <- iris$Sepal.Width
#' # Scatter plot
#' require("rgl")
#' rgl_init()
#' rgl.spheres(x, y, z, r = 0.2, color = "grey")
#' }
#' @export
rgl_init <- function(new.device = FALSE, bg = "white", width = 640) {

  if (!requireNamespace("rgl", quietly = TRUE))
    stop("rgl package needed for this function to work. Please install it.")

  if( new.device | rgl::rgl.cur() == 0 ) {
    rgl::rgl.open()
    rgl::par3d(windowRect = 50 + c( 0, 0, width, width ) )
    rgl::rgl.bg(color = bg )
  }
  rgl::rgl.clear(type = c("shapes", "bboxdeco"))
  rgl::rgl.viewpoint(theta = 45, phi = 30, fov= 60)
}


R/s3d_addgrids.R
#' Add grids to a scatterplot3d
#'
#' @description This function adds grids on an existing
#'  plot created using the package scatterplot3d
#' @param x,y,z numeric vectors specifying the x, y, z coordinates of points.
#'  x can be a matrix or a data frame containing 3 columns corresponding to
#'  the x, y and z coordinates. In this case the arguments y and z are optional
#' @param grid specifies the facet(s) of the plot on which grids should be drawn.
#'  Possible values are the combination of "xy", "xz" or "yz".
#'  Example: grid = c("xy", "yz"). The default value is TRUE to add grids only on xy facet.
#' @param col.grid,lty.grid color and line type to be used for grids
#' @param lab a numerical vector of the form c(x, y, len).
#'  The values of x and y give the (approximate) number of tickmarks on the x and y axes.
#' @param lab.z the same as lab, but for z axis
#' @param scale.y scale of y axis related to x- and z axis
#' @param angle angle between x and y axis
#' @param xlim,ylim,zlim the x, y and z limits (min, max) of the plot.
#'
#' @note
#' Users who want to extend an existing scatterplot3d graphic with the
#'  function s3d_addgrids, should consider to set the arguments scale.y, angle, ...,
#'  to the value used in scatterplot3d.
#' @examples
#' library("scatterplot3d")
#' data(iris)
#'
#' # Basic 3d graphics
#' scatterplot3d(iris[,1:3], pch = 16)
#'
#' # Remove box and add grids
#' scatterplot3d(iris[, 1:3], pch = 16, grid = TRUE, box = FALSE)
#' s3d_addgrids(iris[, 1:3], grid = c("xy", "xz", "yz"))
#'
#' @export
s3d_addgrids <- function(x, y=NULL, z=NULL, grid = TRUE,
                    col.grid = "grey", lty.grid = par("lty"),
                    lab = par("lab"), lab.z = mean(lab[1:2]),
                    scale.y = 1, angle = 40,
                    xlim=NULL, ylim=NULL, zlim=NULL){


  if(inherits(x, c("matrix", "data.frame"))){
    x <- as.data.frame(x)
    y <- unlist(x[,2])
    z <- unlist(x[,3])
    x <- unlist(x[,1])
  }

  p.lab <- par("lab")

  angle <- (angle%%360)/90
  yz.f <- scale.y * abs(if (angle < 1) angle else if (angle >3) angle - 4 else 2 - angle)
  yx.f <- scale.y * (if (angle < 2) 1 - angle else angle - 3)


  # x axis range
  x.range <- range(x[is.finite(x)], xlim)
  x.prty <- pretty(x.range, n = lab[1], min.n = max(1, min(0.5 *lab[1], p.lab[1])))
  x.scal <- round(diff(x.prty[1:2]), digits = 12)
  x <- x/x.scal
  x.range <- range(x.prty)/x.scal
  x.max <- ceiling(x.range[2])
  x.min <- floor(x.range[1])
  if (!is.null(xlim)) {
    x.max <- max(x.max, ceiling(xlim[2]/x.scal))
    x.min <- min(x.min, floor(xlim[1]/x.scal))
  }
  x.range <- range(x.min, x.max)

  # y axis range
  y.range <- range(y[is.finite(y)], ylim)
  y.prty <- pretty(y.range, n = lab[2], min.n = max(1, min(0.5 *lab[2], p.lab[2])))
  y.scal <- round(diff(y.prty[1:2]), digits = 12)
  y.add <- min(y.prty)
  y <- (y - y.add)/y.scal
  y.max <- (max(y.prty) - y.add)/y.scal
  if (!is.null(ylim))
    y.max <- max(y.max, ceiling((ylim[2] - y.add)/y.scal))

  # Z axis range
  z.range <- range(z[is.finite(z)], zlim)
  z.prty <- pretty(z.range, n = lab.z, min.n = max(1, min(0.5 *lab.z, p.lab[2])))
  z.scal <- round(diff(z.prty[1:2]), digits = 12)
  z <- z/z.scal
  z.range <- range(z.prty)/z.scal
  z.max <- ceiling(z.range[2])
  z.min <- floor(z.range[1])
  if (!is.null(zlim)) {
    z.max <- max(z.max, ceiling(zlim[2]/z.scal))
    z.min <- min(z.min, floor(zlim[1]/z.scal))
  }
  z.range <- range(z.min, z.max)

  # Add grid
  if ("xy" %in% grid || grid == TRUE) {
    i <- x.min:x.max
    segments(i, z.min, i + (yx.f * y.max), yz.f * y.max +
               z.min, col = col.grid, lty = lty.grid)
    i <- 0:y.max
    segments(x.min + (i * yx.f), i * yz.f + z.min, x.max +
               (i * yx.f), i * yz.f + z.min, col = col.grid, lty = lty.grid)
  }

  if ("xz" %in% grid) {
    i <- x.min:x.max
    segments(i + (yx.f * y.max), yz.f * y.max + z.min,
             i + (yx.f * y.max), yz.f * y.max + z.max,
             col = col.grid, lty = lty.grid)
    temp <- yx.f * y.max
    temp1 <- yz.f * y.max
    i <- z.min:z.max
    segments(x.min + temp,temp1 + i,
             x.max + temp,temp1 + i , col = col.grid, lty = lty.grid)

  }

  if ("yz" %in% grid) {
    i <- 0:y.max
    segments(x.min + (i * yx.f), i * yz.f + z.min,
             x.min + (i * yx.f) ,i * yz.f + z.max,
             col = col.grid, lty = lty.grid)
    temp <- yx.f * y.max
    temp1 <- yz.f * y.max
    i <- z.min:z.max
    segments(x.min + temp,temp1 + i,
             x.min, i , col = col.grid, lty = lty.grid)
    }

}


R/graph3d-package.R
#' Build and customize 3D graphs in R
#' @name graph3d
#' @aliases graph3d
#' @docType package
#' @description Utilities for easily building and customizing 3D graphs in R.
#' @examples
#' \donttest{
#' # Load dataset
#' data(iris)
#' x <- iris$Sepal.Length
#' y <- iris$Petal.Length
#' z <- iris$Sepal.Width
#'
#' library("rgl")
#'
#' # Quick scatter plots
#' # ++++++++++++++++
#' rgl_scatter(iris, groups = iris$Species)
#' # Scale data
#' rgl_scatter(iris, groups = iris$Species, data.scale = TRUE)
#' # Remove the bounding box
#' rgl_scatter(iris, groups = iris$Species, show.bbox = FALSE)
#' # Add ellipses
#' rgl_scatter(iris, groups = iris$Species,
#'    show.bbox = FALSE, add.ellipse = TRUE)
#'
#' # Build RGL scatter plot
#' # +++++++++++++++++++
#' rgl_init()
#' rgl.spheres(x, y, z, r = 0.2, color = get_colors(iris$Species))
#' rgl_add_axes(x, y, z, show.bbox = TRUE)
#' rgl_add_ellipses(x, y, z, groups = iris$Species)
#' aspect3d(1,1,1)
#' }
#'
NULL


R/rgl_scatter.R
#' @include rgl_init.R
NULL
#' RGL scatter plot
#'
#' @description
#' RGL scatter plot. \cr\cr
#' \itemize{
#' \item rgl_scatter(): quick scatter plot using rgl
#' \item rgl_add_axes(): Add axes to an RGL plot
#' \item get_colors(): Get colors for the different levels of a factor variable.
#' \item rgl_add_ellipses(): Add an ellipse of concentration.
#' \item rgl_add_linear_planes(): Add regression planes
#' \item rgl_group_labels(): Show the label of each group
#' }
#'
#' @param x,y,z the coordinates of the points to draw.
#'  Both y and z can be left at NULL. x can be a data frame or a formula
#'  zvar ~ xvar + yvar
#' @param axis.col colors for the 3 axes.
#' @param show.axis.lab If TRUE show axis labels.
#' @param xlab,ylab,zlab labels to be used for x, y, and z axes, respectively.
#' @param show.plane If TRUE, show axis planes at y = 0.
#' @param plane.col the color to be used for plane. Default is "grey".
#' @param show.bbox add the bounding box decoration.
#' @param bbox.col the bounding box colors. The first color is the
#' the background color; the second color is the color of tick marks
#' @param show.ticks if TRUE, show ticks
#' @param xlim,ylim,zlim the limits for x, y and z axes, respectively
#' @examples
#'\donttest{
#' # Load dataset
#' data(iris)
#' x <- iris$Sepal.Length
#' y <- iris$Petal.Length
#' z <- iris$Sepal.Width
#'
#' library("rgl")
#'
#' # Quick scatter plots
#' rgl_scatter(iris, groups = iris$Species)
#' # Scale data
#' rgl_scatter(iris, groups = iris$Species, data.scale = TRUE)
#' # Remove the bounding box
#' rgl_scatter(iris, groups = iris$Species, show.bbox = FALSE)
#' # Add ellipses
#' rgl_scatter(iris, groups = iris$Species,
#'    show.bbox = FALSE, add.ellipse = TRUE)
#'
#' # Scatter plot with axes
#' # +++++++++++++++++
#' rgl_init()
#' rgl.spheres(x, y, z, r = 0.2, color = "yellow")
#' rgl_add_axes(x, y, z)
#'
#' # Change the color by groups
#' # +++++++++++++++++
#'  rgl_init()
#'  rgl.spheres(x, y, z, r = 0.2, color = get_colors(iris$Species))
#'  rgl_add_axes(x, y, z, show.bbox = TRUE)
#'  aspect3d(1,1,1)
#'
#'  # Add ellipses
#'  # ++++++++++++++++++
#'  rgl_init()
#'  rgl.spheres(x, y, z, r = 0.2, color = "#D95F02")
#'  rgl_add_axes(x, y, z, show.bbox = TRUE)
#'  rgl_add_ellipses(x, y, z, ellipse.col = "#D95F02")
#'  aspect3d(1,1,1)
#'
#'  # Regression planes
#'  # +++++++++++++++++++
#'  rgl_init()
#'  rgl.spheres(x, y, z, r = 0.2,
#'      color = get_colors(iris$Species, c("red", "blue", "green")))
#'  rgl_add_axes(x, y, z, show.bbox = TRUE)
#'  rgl_add_linear_planes(x, y, z, groups = iris$Species,
#'      surface.col = c("red", "blue", "green"))
#'  aspect3d(1,1,1)
#'
#'}
#'
#' @name rgl_scatter
#' @rdname rgl_scatter
#' @param sphere.size the size of the spheres
#' @param point.col the color to be used for the points when groups = NULL
#' @param add.ellipse logical value. If TRUE, an ellipse is added around points
#' @param add.reg.planes logical value. If TRUE add regression planes
#' @param show.group.labels logical value. If TRUE, group labels are shown
#' @param data.scale logical value. If TRUE, the coordinates (x, y, z) is scaled
#'  so that the minimum of each axis = 0 and the maximum = 1
#' @param ... other arguments to be passed to the function rgl_add_axis(),
#' rgl_add_ellipses(), rgl_add_linear_planes() or rgl_group_labels().
#' @export
rgl_scatter <- function(x, y = NULL, z = NULL, groups = NULL, sphere.size = 1,
                        point.col = "grey", group.col = NULL,
                        add.ellipse = FALSE, add.reg.planes = FALSE,
                        show.group.labels = TRUE, data.scale = FALSE,
                        show.bbox = TRUE,
                        ...){

  .check_rgl()
  # Environment setup
  rgl_init()
  # suspend updating while making multiple changes to the scene
  rgl::par3d(skipRedraw= TRUE )

  # Coordinates
  xyz <- grDevices::xyz.coords(x, y, z)
  x <- xyz$x
  y <- xyz$y
  z <- xyz$z

  # Clean the data
  valid <- if (is.null(groups)) complete.cases(x, y, z)
  else complete.cases(x, y, z, groups)
  x <- x[valid]
  y <- y[valid]
  z <- z[valid]

  if(data.scale){
    x <-(x-min(x))/(max(x)-min(x))
    y <-(y-min(y))/(max(y)-min(y))
    z <-(z-min(z))/(max(z)-min(z))
  }

  # Compute spheres size
  if(data.scale) size <- sphere.size*((100/length(x))^(1/3)) * 0.015
  else size <- sphere.size*((100/length(x))^(1/3))* 0.1

  # color
  color <- point.col
  if(!is.null(groups)){
    groups <- groups[valid]
    color <- get_colors(groups, group.col)
  }

  # Plot the data
  if (size > 0.01) rgl::rgl.spheres(x, y, z, color = color, radius = size)
  else rgl::rgl.points(x, y, z, color = color)
  rgl_add_axes(x, y, z, show.bbox = show.bbox,...) # Add axes

  # Add ellipse
  ellipse.col <- point.col[1]
  if(!is.null(groups)) ellipse.col <- group.col
  if(add.ellipse) rgl_add_ellipses(x, y, z, groups = groups, ellipse.col = ellipse.col, ...)

  # Add regression planes
  surface.col <- point.col[1]
  if(!is.null(groups)) surface.col <- group.col
  if(add.reg.planes) rgl_add_linear_planes(x, y, z, groups = groups, surface.col = surface.col, ...)

  # Add group labels
  if(!is.null(groups) & show.group.labels)
    rgl_group_labels(x, y, z, groups = groups, group.col = group.col)


  # Finalize the plot
  rgl::aspect3d( c( 1, 1, 1 ) )
  rgl::par3d(skipRedraw = FALSE)
  rgl::rgl.bringtotop()
}


#' @rdname rgl_scatter
#' @export
rgl_add_axes <- function(x, y = NULL, z = NULL, axis.col = c("grey", "grey", "grey"),
                         show.axis.lab = FALSE, xlab = NULL, ylab = NULL, zlab = NULL,
                         show.plane = TRUE, plane.col = "grey",
                         show.bbox = TRUE, bbox.col = c("#333377","black"),
                         show.ticks = FALSE,
                         xlim = NULL, ylim = NULL, zlim = NULL, ...)
{

  .check_rgl()

  # Coordinates
  xyz <- grDevices::xyz.coords(x, y, z)
  x <- xyz$x
  y <- xyz$y
  z <- xyz$z

  # Axis limits
  if(is.null(xlim)) xlim <- c(-max(abs(x)), max(abs(x))) * 1.1
  if(is.null(ylim)) ylim <- c(-max(abs(y)), max(abs(y))) * 1.1
  if(is.null(zlim)) zlim <- c(-max(abs(z)), max(abs(z))) * 1.1

  # Add axis
  if(length(axis.col) < 3) axis.col <- rep(axis.col, 3)
  rgl::rgl.lines(xlim, c(0, 0), c(0, 0), color = axis.col[1])
  rgl::rgl.lines(c(0,0), ylim, c(0, 0), color = axis.col[2])
  rgl::rgl.lines(c(0,0), c(0, 0), zlim, color = axis.col[3])

  # Add a point at the end of each axes to specify the direction
  rgl::rgl.points(xlim[2],  0, 0, color = axis.col[1], size = 3)
  rgl::rgl.points(0, ylim[2], 0, color = axis.col[2], size = 3)
  rgl::rgl.points(0, 0, zlim[2], color = axis.col[3], size = 3)

  # Add labels
  if(show.axis.lab){
    if(is.null(xlab)) xlab = deparse(substitute(x))
    if(is.null(ylab)) ylab = deparse(substitute(y))
    if(is.null(zlab)) zlab = deparse(substitute(z))
    rgl::rgl.texts(xlim[2],  0, 0, xlab, color = axis.col[1], adj = c(0.5, -0.8), size = 2)
    rgl::rgl.texts(0, ylim[2], 0, ylab, color = axis.col[2],  adj = c(0.5, -0.8), size = 2)
    rgl::rgl.texts(0, 0, zlim[2], zlab,  color = axis.col[3], adj = c(0.5, 0.8), size = 2)
  }

  # Add plane
  if(show.plane) {
    xlim <- xlim/1.1
    zlim <- zlim/1.1

    rgl::rgl.quads(
      rbind( c( xlim[1], 0, zlim[1]), c( xlim[1], 0, zlim[2] ),
             c( xlim[2], 0, zlim[2] ), c( xlim[2], 0, zlim[1] )),
      col= plane.col, alpha= 0.4)
  }

  # Add bounding box decoration
  if(show.bbox){
    rgl::rgl.bbox(color=c(bbox.col[1],bbox.col[2]), alpha = 0.5,
             emission=bbox.col[1], specular=bbox.col[1], shininess=5,
             xlen = 3, ylen = 3, zlen = 3)
  }

  # Show ticks
  if(show.ticks){
    rgl::axis3d('x',pos=c( NA, 0, 0 ), col = axis.col[1])
    rgl::axis3d('y',pos=c( 0, NA, 0 ), col = axis.col[2])
    rgl::axis3d('z',pos=c( 0, 0, NA ), col = axis.col[3])
  }

  rgl::rgl.bringtotop()
}






#' @param group.col a vector containing the names of colors to be used for each group
#' @rdname rgl_scatter
#' @export
get_colors <- function(groups, group.col = palette()[-1]){
  if(is.null(group.col)) group.col <- palette()[-1]
  groups <- as.factor(groups)
  ngrps <- length(levels(groups))
  if(ngrps > length(group.col))
    group.col <- rep(group.col, ngrps)
  color <- group.col[as.numeric(groups)]
  names(color) <- as.vector(groups)
  return(color)
}


#' @param groups a factor variable specifing the groups of the observations
#' @param conf.level The confidence level of a simultaneous confidence region.
#'  This is used to control the size of the ellipsoid.
#' @param ellipse.col a vector of color(s) to be used for the ellipse(s)
#' @param ellipse.fill a logical value. If TRUE, fills the plotted ellipse(s) with ellipse.col
#' @param ellipse.grid a logical value. If TRUE, adds grid lines on the ellipse(s)
#' @rdname rgl_scatter
#' @export
rgl_add_ellipses <- function(x, y = NULL, z = NULL, groups = NULL, conf.level = 0.95,
                             ellipse.col = NULL, ellipse.fill = TRUE, ellipse.grid = FALSE, ...)
{
  .check_rgl()

  # Coordinates
  xyz <- grDevices::xyz.coords(x, y, z)
  x <- xyz$x
  y <- xyz$y
  z <- xyz$z
  if(is.null(groups)) groups <- rep(1, length(x))

  # plot
  groups <- as.factor(groups)
  levs <- levels(groups)

  if(is.null(ellipse.col)) ellipse.col<- palette()[-1]
  if(length(levs)> length(ellipse.col)) ellipse.col <- rep(ellipse.col, length(levs))
  for (i in 1:length(levs)) {
    group <- levs[i]
    selected <- groups == group
    xx <- x[selected]
    yy <- y[selected]
    zz <- z[selected]
    ellips <- rgl::ellipse3d(cov(cbind(xx,yy,zz)),
                        centre=c(mean(xx), mean(yy), mean(zz)), level = conf.level)

    if(ellipse.fill) rgl::shade3d(ellips, col = ellipse.col[i],
                             alpha = 0.1, lit = FALSE)
    if(ellipse.grid) rgl::wire3d(ellips, col = ellipse.col[i],  lit = FALSE)

  }

  rgl::rgl.bringtotop()
}


#' @param surface.col color to be used for regression plane surfaces.
#' @param grid.col color for grids in the regression planes
#' @param plane.fill logical value; if TRUE the regression planes are filled
#' @param plane.grid logical value. If TRUE grids are shown
#' @param droplines logical values. If TRUE, droplines are shown from regression surfaces to the points.
#' @param smooth logical value. If TRUE, smoothed regression planes are drawn.
#' @rdname rgl_scatter
#' @export
rgl_add_linear_planes <- function(x, y = NULL, z=NULL, groups = NULL, surface.col = "steelblue",
                              grid.col = "black", plane.fill = TRUE, plane.grid = FALSE,
                              droplines = TRUE, smooth = FALSE, ...){

  .check_rgl()
  # Coordinates
  xyz <- grDevices::xyz.coords(x, y, z)
  x <- xyz$x
  y <- xyz$y
  z <- xyz$z


  if(is.null(groups))
    .add_rgl_planes(x, y, z, surface.col = surface.col, grid.col = grid.col,
                    plane.fill = plane.fill, plane.grid = plane.grid, droplines = droplines, smooth= smooth)
  else{

    groups <- as.factor(groups)
    levs <- levels(groups)

    if(is.null(surface.col)) surface.col<- palette()[-1]
    if(length(levs) > length(surface.col)) surface.col <- rep(surface.col, length(levs))
    if(!plane.fill) grid.col = surface.col

    for (i in 1:length(levs)) {
      group <- levs[i]
      selected <- groups == group
      xx <- x[selected]
      yy <- y[selected]
      zz <- z[selected]

      .add_rgl_planes(xx, yy, zz, surface.col = surface.col[i], grid.col = grid.col,
                      plane.fill = plane.fill, plane.grid = plane.grid, droplines = droplines, smooth = smooth)


    }

  }

  rgl::rgl.bringtotop()
}

# helper function to addrgl planes
.add_rgl_planes <- function(x, y, z, surface.col = "steelblue",
                            grid.col = "black", plane.fill = TRUE, plane.grid = FALSE,
                            droplines = TRUE, smooth = FALSE){

  # Compute the linear regression (y = ax + bz + d)
  if(smooth) fit = mgcv::gam(y ~ s(x, z))
  else fit <- lm(y ~ x + z)
  # predict values on regular xz grid
  grid.lines = 26
  x.pred <- seq(min(x), max(x), length.out = grid.lines)
  z.pred <- seq(min(z), max(z), length.out = grid.lines)
  xz <- expand.grid( x = x.pred, z = z.pred)

  y.pred <- matrix(predict(fit, newdata = xz),
                   nrow = grid.lines, ncol = grid.lines)

  if (plane.fill)
    rgl::rgl.surface(x.pred, z.pred, y.pred, color = surface.col,
                alpha = 0.5, lit = FALSE)
  if(plane.grid)
    rgl::rgl.surface(x.pred, z.pred, y.pred, color = grid.col[1],
                alpha = 0.5, lit = FALSE, front = "lines", back = "lines")
  if(droplines){
    # fitted points for droplines to surface
    fitted <- fitted(fit)
    rgl::rgl.lines(as.vector(rbind(x, x)), as.vector(rbind(y, fitted)),
              as.vector(rbind(z, z)), color = surface.col[1])
  }


}

#' @param cex the size of text labels
#' @rdname rgl_scatter
#' @export
rgl_group_labels <- function(x, y = NULL, z = NULL,
                             groups, group.col = "black", cex = 2, ...)
{
  # Coordinates
  xyz <- xyz.coords(x, y, z)
  x <- xyz$x
  y <- xyz$y
  z <- xyz$z

  groups <- as.factor(groups)
  levs <- levels(groups)

  if(is.null(group.col)) group.col<- palette()[-1]
  if(length(levs) > length(group.col)) group.col <- rep(group.col, length(levs))

  for (i in 1:length(levs)) {
    group <- levs[i]
    selected <- groups == group
    xx <- x[selected]
    yy <- y[selected]
    zz <- z[selected]
    rgl::texts3d(mean(xx),mean(yy), mean(zz), group, col= group.col[i], cex = cex, ...)
  }

}

.check_rgl <- function(){
  if (!requireNamespace("rgl", quietly = TRUE))
    stop("rgl package needed for this function to work. Please install it.")
}
